src/aggregation/metric/cardinality.rs

```diff
@@ -0,0 +1,381 @@
+    pub fn from_field_name(field_name: String) -> Self {
+        Self {
+            field: field_name,
+            missing: None,
+        }
+    }
+    /// Returns the field name the aggregation is computed on.
+    pub fn field_name(&self) -> &str {
+        &self.field
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub(crate) struct SegmentCardinalityCollector {
+    cardinality: CardinalityCollector,
+    entries: FxHashSet<u64>,
+    column_type: ColumnType,
+    accessor_idx: usize,
+    missing: Option<Key>,
+}
+
+impl SegmentCardinalityCollector {
+    pub fn from_req(column_type: ColumnType, accessor_idx: usize, missing: &Option<Key>) -> Self {
+        Self {
+            cardinality: CardinalityCollector::new(),
+            entries: Default::default(),
+            column_type,
+            accessor_idx,
+            missing: missing.clone(),
+        }
+    }
+
+    fn collect_block_with_field(
+        &mut self,
+        docs: &[crate::DocId],
+        agg_accessor: &mut AggregationWithAccessor,
+    ) {
+        if let Some(missing) = agg_accessor.missing_value_for_accessor {
+            agg_accessor.column_block_accessor.fetch_block_with_missing(
+                docs,
+                &agg_accessor.accessor,
+                missing,
+            );
+        } else {
+            agg_accessor
+                .column_block_accessor
+                .fetch_block(docs, &agg_accessor.accessor);
+        }
+    }
+
+    fn into_intermediate_metric_result(
+        mut self,
+        agg_with_accessor: &AggregationWithAccessor,
+    ) -> crate::Result<IntermediateMetricResult> {
+        if self.column_type == ColumnType::Str {
+            let mut buffer = String::new();
+            let entries: Vec<u64> = self.entries.into_iter().collect();
+            let term_dict = agg_with_accessor.str_dict_column.as_ref().cloned().unwrap();
+            for term_id in entries {
+                if term_id == u64::MAX {
+                    let missing_key = self
+                        .missing
+                        .as_ref()
+                        .expect("Found placeholder term_id but `missing` is None");
+                    match missing_key {
+                        Key::Str(missing) => {
+                            self.cardinality.sketch.insert_any(&missing);
+                        }
+                        Key::F64(val) => {
+                            let val = f64_to_u64(*val);
+                            self.cardinality.sketch.insert_any(&val);
+                        }
+                    }
+                } else {
+                    if !term_dict.ord_to_str(term_id, &mut buffer)? {
+                        return Err(TantivyError::InternalError(format!(
+                            "Couldn't find term_id {term_id} in dict"
+                        )));
+                    }
+                    self.cardinality.sketch.insert_any(&buffer);
```

Comment:
I think the result woudl be more accurate if we pushed a type discriminant in the sketch.

The most efficient way to do this would be to use a salted hasher.

Reply:
Not sure I understand here. Is that about fields that hold multiple value types?

Reply:
done. Test case added as well.

Reply:
it's about different types, we support mixed types on the same column name. Internally they we differentiate them via `(ColumnType, Column)`.
e.g. the bytes of the number `[1u8,2,3,4,5,6,7,8]` are not equal to the bytes of a string `[1u8,2,3,4,5,6,7,8]`. It's not a problem during collection, because we generate two collectors, one for numbers and one for strings. But it may cause collisions during merging.